extension/sigv4authextension/extension_test.go

```diff
@@ -0,0 +1,127 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sigv4authextension
+
+import (
+	"net/http"
+	"testing"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/credentials"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/zap"
+)
+
+func TestNewSigv4Extension(t *testing.T) {
+	cfg := &Config{Region: "region", Service: "service", RoleArn: "rolearn"}
+
+	sa := newSigv4Extension(cfg, "awsSDKInfo", zap.NewNop())
+	assert.Equal(t, cfg.Region, sa.cfg.Region)
+	assert.Equal(t, cfg.Service, sa.cfg.Service)
+	assert.Equal(t, cfg.RoleArn, sa.cfg.RoleArn)
+}
+
+func TestRoundTripper(t *testing.T) {
+	awsCreds := fetchMockCredentials()
+	base := (http.RoundTripper)(http.DefaultTransport.(*http.Transport).Clone())
+	awsSDKInfo := "awsSDKInfo"
+	cfg := &Config{Region: "region", Service: "service", RoleArn: "rolearn", creds: awsCreds}
+
+	sa:= newSigv4Extension(cfg, awsSDKInfo, zap.NewNop())
+	assert.NotNil(t, sa)
+
+	rt, err := sa.RoundTripper(base)
+	assert.Nil(t, err)
+
+	si := rt.(*SigningRoundTripper)
+	assert.Equal(t, base, si.transport)
+	assert.Equal(t, cfg.Region, si.region)
+	assert.Equal(t, cfg.Service, si.service)
+	assert.Equal(t, awsSDKInfo, si.awsSDKInfo)
+
+}
+
+func TestPerRPCCredentials(t *testing.T) {
+	cfg := &Config{Region: "region", Service: "service", RoleArn: "rolearn"}
+	sa := newSigv4Extension(cfg, "", zap.NewNop())
+
+	rpc, err := sa.PerRPCCredentials()
+	assert.Nil(t, rpc)
+	assert.Error(t, err)
+}
+
+func TestGetCredsFromConfig(t *testing.T) {
+	awsCreds := fetchMockCredentials()
+	t.Setenv("AWS_ACCESS_KEY_ID", awsCreds.AccessKeyID)
```

Comment:
I've added back in environment variables into the test for a success case, since according to this [guide](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-quickstart.html#cli-configure-quickstart-precedence), environment variables have the highest precedence for credentials (besides CLI options). Without it, we would look at the `~/.aws` files, and then at container and instance profile credentials. If we don't set environment variables here, then it will eventually error when it reaches the end and tries to use instance profile credentials. I've added in a test for missing creds, and it errors here at the last step of precedence for credentials. It also makes a request to aws, which we want to avoid, so not sure if having this test is ok or not, or if there's another way to go about this test.

Reply:
Thanks for looking into this - I guess it is hitting IMDS to try to fetch a cred for an EC2 machine. Since that IP will almost always be unresolvable it's a bit better than hitting the STS endpoint which is always available so maybe we can live with it. Could you try if disabling IMDS helps? Can follow up in another PR though

`AWS_EC2_METADATA_DISABLED`
https://github.com/aws/aws-sdk-go-v2/blob/c214cb61990441aa165e216a3f7e845c50d21939/config/load_options.go#L169